### DISCUSSION TITLE
Upgraded from beta 2 to latest and return type on findMany doesn't include types in the `include : { }` prop

### DISCUSSION THREAD

**Discussion by webnoob** [prisma/prisma#3904] [Q&A] [answered]
Before, I could do:

```ts
    const posts = await this.dbService.messagePost.findMany({
      include: {
        user: {
          select: {
            id: true
          }
        }
      })
    })

    for (const post of posts) {
      console.log(post.user) // THIS NOW HAS A TYPE ERROR
    }
```

Before, the compile knew `user` existed on `post` but now it doesn't. Has something changed that would cause this?

The error:  `Property 'user' does not exist on type 'MessagePost & {}'.`

**Comment by webnoob** [comment, marked as answer]
On reflection, this seems to be an issue with some custom code which I use to get the includes (the example above is simplified).
I have a function which I use to get common includes within my service with a signature like this: 

```
public getStandardPostIncludes (extend?: MessagePostInclude): MessagePostInclude {
  return {
    someField: true,
    ...extend
  }
}
```

For some reason, despite this having the same return type as `include` in findMany, it won't recognise the types when I use it.

**Reply by ryands17** [reply]
I'm on version _2.8.1_ of `@prisma/client` and `@prisma/cli` and used this mode for simplicity.

```prisma
model User {
  id    Int    @id @default(autoincrement())
  name  String
  posts Post[]
}

model Post {
  id     Int    @id @default(autoincrement())
  title  String
  user   User   @relation(fields: [userId], references: [id])
  userId Int
}
```

And this gives the correct type:

```ts
const posts = await prisma.post.findMany({
    include: {
      user: {
        select: {
          id: true,
        },
      },
    },
  })

  for (const post of posts) {
    console.log(post.user)
  }
```

![image](https://user-images.githubusercontent.com/19697099/95860432-aaa5d580-0d7d-11eb-9e21-51ba7ea19017.png)

Could you check your Prisma Client and CLI version and run `prisma generate` again to check?

**Reply by webnoob** [reply]
Yes, sorry. I dug deeper and it seems its an issue with getting the includes from a function. Can you see my edit and let me know your thoughts?

**Reply by webnoob** [reply]
My actual query in line with my edits would be:

```
const posts = await this.dbService.messagePost.findMany({
      include: this.getStandardThreadIncludes({
        user: {
          select: {
            id: true
          }
        })
      })
    })

    for (const post of posts) {
      console.log(post.user) // THIS NOW HAS A TYPE ERROR
    }
```

**Reply by ryands17** [reply]
Ahh. You are dynamically providing values to the `select` field which is causing the error. Type it statically and you will get it fine.

This is more of a TS thing than a Prisma thing as TypeScript doesn't know which fields you are providing in `select` runtime. So most probably you would need to specify something that tells the function what it's returning which might help TS give the correct type.

Hope that makes sense :)

**Reply by webnoob** [reply]
Hmm, well it's been working up until this point without issue. 

Has the generation of the `MessagePostGetPayload` type been changed between the versions?

**Reply by webnoob** [reply]
Also, the return type of the function is static. It returns a `MessagePostInclude` so I would have thought it could determine it regardless 🤔

**Reply by ryands17** [reply]
Could you provide the body somewhat of the function `getStandardThreadIncludes`?

> Has the generation of the MessagePostGetPayload type been changed between the versions?

Quite possibly. As you were using `beta.2` which is quite old, it is possible.

**Reply by ryands17** [reply]
I can check it by implementing it on my end

**Reply by webnoob** [reply]
Here it is:

```
public getStandardPostIncludes (extend?: MessagePostInclude): MessagePostInclude {
    return {
      user: { // The user sent it
        select: {
          id: true,
          image: true,
          emailAddress: true,
          name: true
        }
      },
      ...extend
    }
}

**Reply by webnoob** [reply]
In my code this query would be used:

```
const posts = await this.dbService.messagePost.findMany({
      include: this.getStandardPostIncludes()
})
```

Sorry for formatting, doing it in this editor.

**Reply by webnoob** [reply]
The general idea being the function is used to get the "general includes" I use in a lot of queries but I can add extras in if needs be.

**Reply by webnoob** [reply]
Ok, some progress, if I change the signature of my function to this, it works as expected:

```
public getStandardPostIncludes (extend?: MessagePostInclude): MessagePostInclude & {
    user: boolean | UserArgs
} {
  // return something
}
```
I guess this makes sense as in the normal declaration it's `user?: boolean | UserArgs` whereas I'm explicitly defining it so need to tell the compile it's there. 
So, for each type I add to my default return object, I'll specify it in the extended return type for that function.

**Reply by ryands17** [reply]
That's because the return type is `PostInclude` (in your case `MessagePostInclude`) TS cannot guess that user is true. The structure of `PostInclude` is something like:

```ts
export type PostInclude = {
  user?: boolean | UserArgs
}
```

Which just says that the user type is optional but doesn't enforce it to be true. You would explicitly need to tell TS that `user` is included.

And yes the above is correct as TS explicitly knows that `user` is true.

**Reply by ryands17** [reply]
> So, for each type I add to my default return object, I'll specify it in the extended return type function

This will only work for default objects and not the ones added dynamically.

**Reply by webnoob** [reply]
Hmm, Ok. I might have to just take each case it comes in that case.
I still find it odd that this is breaking now as the conditional typing was in place before. I might roll back and compare the generated code. I find it odd because user in this case has always been optional so how could the compiler work it out before and not now 🤔

**Reply by webnoob** [reply]
BTW, thanks for your time. I appreciate it.

**Reply by ryands17** [reply]
Let me know if you come to know why it was working previously :)

**Reply by webnoob** [reply]
Ok, I just downgraded to my previous version and got the same errors. I didn't undo _all_ my changes so perhaps there is a TS change that has caused this.

Still, I have it working and the errors were not down to prisma so looks to be a red herring.

Thanks again.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
